include/nighthawk/adaptive_load/metrics_evaluator.h

```diff
@@ -0,0 +1,87 @@
+#include "envoy/config/core/v3/base.pb.h"
+
+#include "nighthawk/adaptive_load/metrics_plugin.h"
+
+#include "external/envoy/source/common/common/logger.h"
+#include "external/envoy/source/common/common/statusor.h"
+#include "external/envoy/source/common/protobuf/protobuf.h"
+
+#include "api/adaptive_load/adaptive_load.pb.h"
+#include "api/adaptive_load/benchmark_result.pb.h"
+#include "api/adaptive_load/metric_spec.pb.h"
+#include "api/client/options.pb.h"
+#include "api/client/output.pb.h"
+#include "api/client/service.pb.h"
+
+#include "absl/container/flat_hash_map.h"
+#include "absl/status/status.h"
+#include "absl/strings/str_join.h"
+
+namespace Nighthawk {
+
+/**
+ * An interface with utilities for translating between metrics definitions, thresholds, scores,
+ * MetricsPlugins, and Nighthawk Service results.
+ */
+class MetricsEvaluator {
+public:
+  virtual ~MetricsEvaluator() = default;
+
+  /**
+   * Given a MetricSpec, obtains a single metric value from the MetricPlugin and optionally scores
+   * it according to a threshold and scoring function.
+   *
+   * @param metric_spec The metric spec identifying the metric by name and plugin name.
+   * @param metrics_plugin An already activated MetricsPlugin used by the metric_spec.
+   * @param threshold_spec A proto describing the threshold and scoring function. Nullptr if the
+   * metric is informational only.
+   *
+   * @return StatusOr<MetricEvaluation> A proto containing the metric value (and its score if a
+   * threshold was specified), or an error status if the metric could not be obtained from the
+   * MetricsPlugin.
+   */
+  virtual absl::StatusOr<nighthawk::adaptive_load::MetricEvaluation>
+  EvaluateMetric(const nighthawk::adaptive_load::MetricSpec& metric_spec,
+                 MetricsPlugin& metrics_plugin,
+                 const nighthawk::adaptive_load::ThresholdSpec* threshold_spec) const PURE;
+
+  /**
+   * Extracts metric descriptors and corresponding thresholds from a top-level adaptive load session
+   * spec to an ordered list and a map. Allows for uniform treatment of scored and informational
+   * metrics.
+   *
+   * @param spec The adaptive load session spec.
+   * @param metric_specs A list to store extracted MetricSpecs in order of definition.
+   * @param threshold_spec_from_metric_spec A map to store each MetricSpec and its threshold if it
+   * had one, or nullptr if it was an informational metric.
+   */
+  virtual void
+  ExtractMetricSpecs(const nighthawk::adaptive_load::AdaptiveLoadSessionSpec& spec,
+                     std::vector<const nighthawk::adaptive_load::MetricSpec*>& metric_specs,
```

Comment:
I wonder if instead of having two output args, a single one consisting of a vector of `std::pair<>`s with a mandatory metric + optional threshold could simplify this interface?

Reply:
Done

Reply:
The reason for the two separate structures is that I couldn't find a standard map that preserves the order of insertion.

Reply:
Made both data structures conts so there is no way for them to get out of sync.